src/circular_buffer.rs

```diff
@@ -0,0 +1,313 @@
+
+#![deny(missing_docs)]
+#![deny(warnings)]
+
+use core::marker::PhantomData;
+
+/// A circular buffer
+pub struct CircularBuffer<T, A>
+where
+    A: AsMut<[T]> + AsRef<[T]>,
+    T: Copy,
+{
+    _marker: PhantomData<[T]>,
+    array: A,
+    index: usize,
+    len: usize,
+}
+
+impl<T, A> CircularBuffer<T, A>
+where
+    A: AsMut<[T]> + AsRef<[T]>,
+    T: Copy,
+{
+    /// Creates a new empty circular buffer using `array` as backup storage.
+    ///
+    /// # Example
+    /// ```
+    /// #![feature(const_fn)]
+    /// use heapless::CircularBuffer;
+    /// const LEN: usize = 16;
+    /// // with let
+    /// let buf_let = CircularBuffer::new([0;LEN]);
+    /// // with const
+    /// const BUF_CONST: CircularBuffer<u8, [u8; LEN]>
+    ///     = CircularBuffer::new([0;LEN]);
+    /// // with static
+    /// static BUF_STATIC: CircularBuffer<u8, [u8; LEN]>
+    ///     = CircularBuffer::new([0;LEN]);
+    /// // with slice
+    /// let mut array = [0_usize; 8];
+    /// let mut buf_slice = CircularBuffer::new(&mut array);
+    /// ```
+    #[inline]
+    pub const fn new(array: A) -> Self {
+        CircularBuffer {
+            _marker: PhantomData,
+            array: array,
+            index: 0,
+            len: 0,
+        }
+    }
+
+    /// Creates a new circular buffer using `array` as backup storage.
+    /// Keeps the values and the length of the array.
+    /// The backup storage is completly used upon creation.
+    /// Insertions will overwrite at the start of the array.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let buf = CircularBuffer::from_array([1,2,3]);
+    /// assert!(buf.is_full());
+    /// ```
+    #[inline]
+    pub fn from_array(array: A) -> Self {
```

Comment:
I'm wondering if this can be named just `from` or if this could be an implementation of the `From` trait. In the latter case, this couldn't be a `const` function.

Reply:
When I think about it, all constructor functions in this crate should be const fn so they can be used with static, so this function should either be removed or re-written to const